stripecardscan/src/main/java/com/stripe/android/stripecardscan/cardscan/CardScanActivity.kt

```diff
@@ -0,0 +1,571 @@
+     */
+    override fun onBackPressed() {
+        runBlocking { scanStat.trackResult("user_canceled") }
+        resultListener.userCanceled(CancellationReason.Back)
+        closeScanner()
+    }
+
+    /**
+     * Handle permission status changes. If the camera permission has been granted, start it. If
+     * not, show a dialog.
+     */
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<out String>,
+        grantResults: IntArray
+    ) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+
+        if (requestCode == PERMISSION_REQUEST_CODE && grantResults.isNotEmpty()) {
+            when (grantResults[0]) {
+                PackageManager.PERMISSION_GRANTED -> {
+                    launch { permissionStat.trackResult("success") }
+                    prepareCamera { onCameraReady() }
+                }
+                else -> {
+                    launch { permissionStat.trackResult("failure") }
+                    userDeniedCameraPermission()
+                }
+            }
+        }
+    }
+
+    /**
+     * Set up viewFinderWindowView and viewFinderBorderView centered with predefined margins
+     */
+    private fun setupViewFinderConstraints() {
+
+        val screenSize = Resources.getSystem().displayMetrics.let {
+            Size(it.widthPixels, it.heightPixels)
+        }
+        val viewFinderMargin = (
+            min(screenSize.width, screenSize.height) *
+                getFloatResource(R.dimen.stripeViewFinderMargin)
+            ).roundToInt()
+
+        listOf(viewBinding.viewFinderWindow, viewBinding.viewFinderBorder).forEach { view ->
+            (view.layoutParams as ViewGroup.MarginLayoutParams)
+                .updateMargins(
+                    viewFinderMargin, viewFinderMargin, viewFinderMargin, viewFinderMargin
+                )
+        }
+
+        viewBinding.viewFinderBackground.setViewFinderRect(viewBinding.viewFinderWindow.asRect())
+    }
+
+    /**
+     * Ensure that the camera permission is available. If so, start the camera. If not, request it.
+     */
+    private fun ensurePermissionAndStartCamera() = when {
+        ContextCompat.checkSelfPermission(
+            this,
+            Manifest.permission.CAMERA,
+        ) == PackageManager.PERMISSION_GRANTED -> {
+            launch { permissionStat.trackResult("success") }
+            prepareCamera { onCameraReady() }
+        }
+        ActivityCompat.shouldShowRequestPermissionRationale(
+            this,
+            Manifest.permission.CAMERA,
+        ) -> showPermissionRationaleDialog()
+        storage.getBoolean(
+            PERMISSION_RATIONALE_SHOWN,
+            false,
+        ) -> showPermissionDeniedDialog()
+        else -> requestCameraPermission()
+    }
+
+    /**
+     * Once the camera stream is available, start processing images.
+     */
+    private fun onCameraStreamAvailable(cameraStream: Flow<CameraPreviewImage<Bitmap>>) {
```

Comment:
is it possible to move the camera functions into the shared parent class as well?